network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -8,38 +10,35 @@ import (
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"github.com/rs/zerolog"
 
-	channels "github.com/onflow/flow-go/engine"
 	"github.com/onflow/flow-go/network"
 
 	"github.com/onflow/flow-go/model/flow"
-	cborcodec "github.com/onflow/flow-go/network/codec/cbor"
 	"github.com/onflow/flow-go/network/message"
 )
 
 // AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
 // is part of the authorized roles list for the channel being communicated on. A node is considered
-// to be authorized to send a message if all of the following are true.
-// 1. The node is authorized.
-// 2. The message type is a known message type (can be decoded with cbor codec).
-// 3. The authorized roles list for the channel contains the senders role.
-// 4. The node is not ejected
-func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
+// to be authorized to send a message if the following are true.
+// 1. The node is staked.
+// 2. The node is not ejected.
+// 3. The message type is a known message type (can be decoded with network codec).
+// 4. The message is authorized to be sent on channel.
+// 4. The sender role is authorized to send message channel.
+// 5. The sender role is authorized to participate on channel.
+func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, c network.Codec, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
```

Comment:
One important thing here:

Since we are now decoding the payload in the topic validator, I think we should update the signature of `MessageValidator` to take in the decoded payload, instead of taking in `message.Message`. The reason is because there may be multiple `MessageValidators` registered to the base Topic Validator: https://github.com/onflow/flow-go/blob/2569fa44b3ca08eec521281fb7c099cb1853664a/network/validator/pubsub/validator.go#L63

We should decode the payload a single time in the TopicValidator, and add the decoded payload to the `ValidatorData`: https://github.com/onflow/flow-go/blob/2569fa44b3ca08eec521281fb7c099cb1853664a/network/validator/pubsub/validator.go#L79-L82

The ValidatorData is later retrieved by the application layer: https://github.com/onflow/flow-go/blob/6c9e6b908ce6aca7be57b3077ef2e9b95d04a8c7/network/p2p/readSubscription.go#L82-L83

From here, we can pass the already decoded payload directly to the callback: https://github.com/onflow/flow-go/blob/6c9e6b908ce6aca7be57b3077ef2e9b95d04a8c7/network/p2p/readSubscription.go#L94

Which calls this function: https://github.com/onflow/flow-go/blob/6c9e6b908ce6aca7be57b3077ef2e9b95d04a8c7/network/p2p/middleware.go#L541-L552

This way, we can avoid decoding the payload a second time at the application layer: https://github.com/onflow/flow-go/blob/6c9e6b908ce6aca7be57b3077ef2e9b95d04a8c7/network/p2p/network.go#L350-L372

Reply:
https://github.com/onflow/flow-go/pull/2491/commits/489ee370357a3be808f998c6ae1f94776c601fab